### PR TITLE
break: ExecuteAsync() under the hood implementation after bumping to Selenium 4.23

### LINKED ISSUE #798: [Bug]: Selenium 4.22 introduces breaking change in ICommandExecutor implementations in Appium
## Description

Updating to Selenium 4.22 introduces a breaking change due to the additional async method ExecuteAsync() ICommandExecutor interface:

https://github.com/SeleniumHQ/selenium/blob/selenium-4.22.0/dotnet/src/webdriver/ICommandExecutor.cs

Selenium devs also rewrote several of their current implementations of Execute() to use ExecuteAsync() (by Task.Run(...)), and I believe that is causing this error when I use Appium 5.0.0 with Selenium 4.22:

'Method 'ExecuteAsync' in type 'OpenQA.Selenium.Appium.Service.AppiumCommandExecutor' from assembly 'Appium.Net, Version=5.0.0.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.'

Reverting back to Selenium 4.21 resolves this.

## Environment

.NET 8.0.6
Appium 5.0.0
Selenium 4.22
Occurs with both AndroidDriver and IOSDriver instantiations

### PR DESCRIPTION
**Pull request by kelmelzer**
## List of changes

This bumps to Selenium 4.23 and fixes the build errors in AppiumDriver and AppiumCommandExecutor to flow through the new ExecuteAsync() methods in Selenium 4.23+.  Fixes #798 
 
## Types of changes

What types of changes are you proposing/introducing to the .NET client?

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] New feature (non-breaking change which adds value to the project)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

### CODE REVIEW

**Comment by KazuCocoa** [comment]
Will this require selenium 3.22+? (Just I don't have much knowledge about deps handling in dotnet lock files, or src/Appium.Net/Appium.Net.csproj handles it)

**Comment by Dor-bl** [comment]
> Will this require selenium 3.22+? (Just I don't have much knowledge about deps handling in dotnet lock files, or src/Appium.Net/Appium.Net.csproj handles it)

@KazuCocoa I assume you meant 4.22+.
Indeed we need to make sure it won't break backward compatibility

**Comment by KazuCocoa** [comment]
Yea, 4.22. If we need to bump the required version, that will also be fine with at least bumping the minor version such as `5.1.0` dotnet driver and changelog etc. A matrix such as https://github.com/appium/python-client?tab=readme-ov-file#compatibility-matrix would help.

**Comment by Dor-bl** [comment]
> Yea, 4.22. If we need to bump the required version, that will also be fine with at least bumping the minor version such as `5.1.0` dotnet driver and changelog etc. A matrix such as https://github.com/appium/python-client?tab=readme-ov-file#compatibility-matrix would help.

At first look, this will be a breaking change since we inherit from `WebDriver` the `ExecuteAsync` method.
A matrix of compatibility is a good idea to have regardless.

**Comment by Dor-bl** [comment]
@nvborisenko your feedback will be appreciated as well :)

File: src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs [hunk 1]
```diff
@@ -54,13 +55,19 @@ internal AppiumCommandExecutor(AppiumLocalService service, TimeSpan timeForTheSe
...
+
+        public async Task<Response> ExecuteAsync(Command commandToExecute)
         {
             Response result = null;
 
             try
             {
                 bool newSession = HandleNewSessionCommand(commandToExecute);
-                result = RealExecutor.Execute(commandToExecute);
+                result = Task.Run(() => RealExecutor.ExecuteAsync(commandToExecute)).GetAwaiter().GetResult();
```
Comment (Dor-bl) [inline src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs:69]: `ConfigureAwait(false)` is generally more efficient because it does not incur the overhead of creating a new thread, and it avoids potential deadlocks or unnecessary blocking caused by capturing the current synchronization context.
You can read more about it in this blog: https://blog.stephencleary.com/2012/02/async-and-await.html

**Review by Dor-bl — CHANGES_REQUESTED** [review]

**Comment by Dor-bl** [comment]
@kelmelzer , the selenium team has released a patch (4.23.0) can you please check if we still need adjustments from our side?

**Comment by kelmelzer** [comment]
Will do when I get a chance! I have been putting out some other fires
lately. :)

**Comment by Dor-bl** [comment]
I hope no fires related to crowdstrike 😅

**Comment by kelmelzer** [comment]
> I hope no fires related to crowdstrike 😅

Definitely related 👎 

Anyways, Appium 5 still doesn't build with Selenium 4.23 due to ICommandExecutor's ExecuteAsync() not being implemented by AppiumCommandExecutor. Adding this simple passthrough fixes:

        public Task<Response> ExecuteAsync(Command commandToExecute) => Task.Run(() => this.Execute(commandToExecute));

**Comment by kelmelzer** [comment]
I (should have) updated my pull request with 4.23 and the small change I mentioned previously.

**Comment by Dor-bl** [comment]
> > I hope no fires related to crowdstrike 😅
> 
> Definitely related 👎 
> 
> Anyways, Appium 5 still doesn't build with Selenium 4.23 due to ICommandExecutor's ExecuteAsync() not being implemented by AppiumCommandExecutor. Adding this simple passthrough fixes:
> 
>         public Task<Response> ExecuteAsync(Command commandToExecute) => Task.Run(() => this.Execute(commandToExecute));

I see the build still fails. Did you checked the changes locally?

**Comment by kelmelzer** [comment]
> > > I hope no fires related to crowdstrike 😅
> > 
> > 
> > Definitely related 👎
> > Anyways, Appium 5 still doesn't build with Selenium 4.23 due to ICommandExecutor's ExecuteAsync() not being implemented by AppiumCommandExecutor. Adding this simple passthrough fixes:
> > ```
> >     public Task<Response> ExecuteAsync(Command commandToExecute) => Task.Run(() => this.Execute(commandToExecute));
> > ```
> 
> I see the build still fails. Did you checked the changes locally?

Let's see what happens with these recent commits...hopefully I got it right.

**Comment by Dor-bl** [comment]
@kelmelzer Please update PR title

File: src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs [hunk 2]
```diff
@@ -79,6 +80,8 @@ public Response Execute(Command commandToExecute)
             }
         }
 
+        public Task<Response> ExecuteAsync(Command commandToExecute) => Task.Run(() => this.Execute(commandToExecute));
```
Comment (Dor-bl) [inline src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs:83]: We should adapt a similar approach to the one the selenium team did.
Meaning, having the Excute method calling the ExcuteAsync and not the other way around.

**Review by Dor-bl — CHANGES_REQUESTED** [review]

File: src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs (hunk 2 above)
Comment (kelmelzer) [reply src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs:83]: ok, no worries, will flip that

File: src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs [hunk 3]
```diff
@@ -53,14 +54,16 @@ internal AppiumCommandExecutor(AppiumLocalService service, TimeSpan timeForTheSe
             ClientConfig = clientConfig;
         }
 
-        public Response Execute(Command commandToExecute)
+        public Response Execute(Command commandToExecute) => this.ExecuteAsync(commandToExecute).GetAwaiter().GetResult();
```
Comment (nvborisenko) [inline src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs:57]: Use `Task.Run(ExecuteAsync(commandToExecute)).GetAwaiter().GetResult()` to avoid deadlocks.

File: src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs [hunk 4]
```diff
@@ -53,14 +54,16 @@ internal AppiumCommandExecutor(AppiumLocalService service, TimeSpan timeForTheSe
...
+
+        public async Task<Response> ExecuteAsync(Command commandToExecute)
         {
             Response result = null;
 
             try
             {
                 bool newSession = HandleNewSessionCommand(commandToExecute);
-                result = RealExecutor.Execute(commandToExecute);
+                result = await RealExecutor.ExecuteAsync(commandToExecute);
```
Comment (nvborisenko) [inline src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs:66]: And `.ConfigureAwait(false)` to be a great library.

**Review by nvborisenko — APPROVED** [review]

**Review by Dor-bl — COMMENTED** [review]
Did you happen to run some tests to make sure everything plays well together?

File: src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs [hunk 5]
```diff
@@ -53,14 +54,16 @@ internal AppiumCommandExecutor(AppiumLocalService service, TimeSpan timeForTheSe
             ClientConfig = clientConfig;
         }
 
-        public Response Execute(Command commandToExecute)
+        public Response Execute(Command commandToExecute) => Task.Run(() => ExecuteAsync(commandToExecute)).GetAwaiter().GetResult();
```
Comment (Dor-bl) [inline src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs:57]: Consider using a block of code.
 ```suggestion
        public Response Execute(Command commandToExecute)
        {
            return Task.Run(() => ExecuteAsync(commandToExecute)).GetAwaiter().GetResult();
        }
```

Comment (Dor-bl) [reply src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs:57]: @kelmelzer let me know if you approve the suggestion. Not critical

Comment (kelmelzer) [reply src/Appium.Net/Appium/Service/AppiumCommandExecutor.cs]: All good. I just like that operator but your suggestion makes sense in the overall scheme of the code :)

File: src/Appium.Net/Appium.Net.csproj [hunk 6]
```diff
@@ -58,9 +58,9 @@
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
-    <PackageReference Include="Selenium.Support" Version="4.21.0" />
-    <PackageReference Include="Selenium.WebDriver" Version="4.21.0" />
-    <PackageReference Include="System.Drawing.Common" Version="8.0.6" />
+    <PackageReference Include="Selenium.Support" Version="4.23.0" />
+    <PackageReference Include="Selenium.WebDriver" Version="4.23.0" />
+    <PackageReference Include="System.Drawing.Common" Version="8.0.7" />
```
Comment (Dor-bl) [inline src/Appium.Net/Appium.Net.csproj:63]: @kelmelzer this should be as a separate PR. After related tests are passing

**Review by Dor-bl — CHANGES_REQUESTED** [review]

**Review by mialeska — APPROVED** [review]

File: src/Appium.Net/Appium.Net.csproj (hunk 6 above)
Comment (Dor-bl) [reply src/Appium.Net/Appium.Net.csproj:63]: @kelmelzer I meant System.Drawing.Common update

Comment (Dor-bl) [inline src/Appium.Net/Appium.Net.csproj:63]:
```suggestion
    <PackageReference Include="System.Drawing.Common" Version="8.0.6" />
```

**Comment by Dor-bl** [comment]
@kelmelzer, Please fix the merge conflicts.

**Comment by kelmelzer** [comment]
> @kelmelzer, Please fix the merge conflicts.

Been out of pocket most of the week. I accepted your change.

**Comment by Dor-bl** [comment]
> > @kelmelzer, Please fix the merge conflicts.
> 
> Been out of pocket most of the week. I accepted your change.

Sorry for the confusion. I updated drawing.common yesterday. Can you please align with the main branch?

**Comment by Dor-bl** [comment]
> > > @kelmelzer, Please fix the merge conflicts.
> > 
> > 
> > Been out of pocket most of the week. I accepted your change.
> 
> Sorry for the confusion. I updated drawing.common yesterday. Can you please align with the main branch?

@kelmelzer, Hope you don't mind but I resolved the conflict myself so we can speed up this PR :)

**Review by Dor-bl — APPROVED** [review]

**Comment by jlipps** [comment]
Hi @kelmelzer , congrats, the Appium project wants to compensate you for this (and perhaps other) contribution(s) this month! Please reply to this comment mentioning me and sharing your [OpenCollective](https://opencollective.com) account name, so that we can initiate payment! Or let me know if you decline to receive compensation via OpenCollective. Thank you!

**Comment by kelmelzer** [comment]
> Hi @kelmelzer , congrats, the Appium project wants to compensate you for this (and perhaps other) contribution(s) this month! Please reply to this comment mentioning me and sharing your [OpenCollective](https://opencollective.com) account name, so that we can initiate payment! Or let me know if you decline to receive compensation via OpenCollective. Thank you!

@jlipps That's wonderful. My account name is kelmelzer there as well.

**Comment by nvborisenko** [comment]
That's great!
